chrome/content/api/FPVS/implementation.js

```diff
@@ -0,0 +1,394 @@
+        // completed. This is gross, but appears to work well enough.
+        // By 2022, the bug is wontfix
+        // This needs to be registered only once, not per window!
+        MailServices.mfn.addListener(
+          this,
+          MailServices.mfn.msgsMoveCopyCompleted |
+          MailServices.mfn.folderMoveCopyCompleted
+        );
+      }
+    }
+
+    remove(callback) {
+      this.off("dragdrop", callback);
+      this.callbackCount--;
+
+      if (this.callbackCount == 0) {
+        log("Remove listener");
+        for (let window of ExtensionSupport.openWindows) {
+          if ([
+            "chrome://messenger/content/messenger.xhtml"
+          ].includes(window.location.href)) {
+            let folderPaneHeader = window.document.getElementById("folderPaneHeader");
+            folderPaneHeader.removeEventListener("mouseout", windowListener);
+            folderPaneHeader.removeEventListener("dragend", windowListener);
+            folderPaneHeader.removeEventListener("dragenter", windowListener);
+
+            let folderTree = window.document.getElementById("folderTree");
+            folderTree.removeEventListener("dragover", windowListener);
+            folderTree.removeEventListener("dragleave", windowListener);
+          }
+        }
+        ExtensionSupport.unregisterWindowListener(this.listenerId);
+
+        // Dragexit and dragdrop don't actually get sent when the user
+        // drops a message into a folder. This is arguably a bug in
+        // Thunderbird (see bz#674807). To work around it, I register a
+        // folder listener to detect when a move or copy is
+        // completed. This is gross, but appears to work well enough.
+        // By 2022, the bug is wontfix
+        // This needs to be registered only once, not per window!
+        MailServices.mfn.removeListener(this);
+      }
+    }
+  };
+
+  var FPVS = class extends ExtensionCommon.ExtensionAPI {
+    // An alternative to defining a constructor here, is to use the onStartup
+    // event. However, this causes the API to be instantiated directly after the
+    // add-on has been loaded, not when the API is first used. Depends on what is
+    // desired.
+    constructor(extension) {
+      super(extension);
+      windowListener = new WindowListener(extension);
+    }
+
+    getAPI(context) {
+      return {
+        FPVS: {
+          initFolderPaneOptionsPopup: async function (windowId) {
+            let mail3Pane = context.extension.windowManager.get(windowId).window;
+            for (; ;) {
+              try {
+                if (mail3Pane.gFolderTreeView.isInited) {
+                  log("treeIsReady", true);
+                  break;
+                }
+              } catch (e) {
+                log("treeIsReady Err", e.message);
+              }
+              await new Promise((resolve) => mail3Pane.setTimeout(resolve, 100));
+              log("treeIsReady", false);
+            }
```

Comment:
I guess the `for (;;) { if (…) { break; } … }` mechanism is not widely known. To provide more maintainability I suggest to convert this to a `do { … } while (…)` or `while (…) { … }` loop.

Reply:
Using a ready state variable now.